File: pine-java/src/test/java/page/liam/pine/StorageModeDispatchTest.java
package page.liam.pine;

import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

/**
 * Pins {@code storage_mode} dispatch to pine-go's rule (issue #179).
 *
 * <p>pine-go's {@code NewFrame} is a {@code switch} on a string-typed
 * {@code StorageMode} with {@code default: newRowFrame}, so exactly one value —
 * the literal {@code "column"} — reaches the column store and everything else
 * reaches the row store. All three runtimes must agree on that, casing included.
 *
 * <p>Before #179 they did not: this runtime matched with
 * {@code equalsIgnoreCase} so {@code "Column"} selected the column store, and
 * pine-cpp's factory fell back to column so a misspelling like {@code "colunm"}
 * did too. One hand-written config therefore selected a different physical store
 * in each runtime.
 *
 * <p>Row and column stores are output-equivalent by design — cross-validate
 * section 4 asserts that — so the divergence never changed a response byte, only
 * the memory and performance profile. That is precisely why it needs a unit test:
 * no end-to-end channel can observe which store was chosen.
 */
class StorageModeDispatchTest {

    private static List<Map<String, Object>> oneItem() {
        List<Map<String, Object>> items = new ArrayList<>();
        Map<String, Object> item = new HashMap<>();
        item.put("id", "a");
        items.add(item);
        return items;
    }

    @Test
    void exactLowercaseColumnSelectsTheColumnStore() {
        Frame f = Frame.create("column", new HashMap<>(), oneItem());
        assertInstanceOf(ColumnFrame.class, f, "exact \"column\" must select the column store");
        assertTrue(f.itemCount() == 1);
    }

    @Test
    void everythingElseSelectsTheRowStore() {
        // Includes the two values that diverged in #179: "Column" (this runtime
        // used equalsIgnoreCase) and "colunm" (pine-cpp fell back to column).
        String[] rowModes = {
            "row", "", "colunm", "Column", "COLUMN", "cOlUmN",
            "column ", " column", "columns", "col", "rows", "unknown",
        };
        for (String mode : rowModes) {
            Frame f = Frame.create(mode, new HashMap<>(), oneItem());
            assertInstanceOf(DataFrame.class, f,
                    "storage_mode \"" + mode + "\" must select the row store");
            assertTrue(f.itemCount() == 1);
        }
    }

    @Test
    void nullStorageModeSelectsTheRowStore() {
        // Reachable when a hand-written config omits the key entirely. Go's
        // switch on the zero value likewise falls to default.
        Frame f = Frame.create(null, new HashMap<>(), oneItem());
        assertInstanceOf(DataFrame.class, f, "absent storage_mode must select the row store");
    }
}

File: pine-go/internal/dataframe/storage_mode_dispatch_test.go
package dataframe

import "testing"

// TestNewFrameDispatchIsExactMatch pins the reference behaviour that pine-java
// and pine-cpp were aligned to in issue #179.
//
// NewFrame switches on a string-typed StorageMode with `default: newRowFrame`,
// so exactly one value — the literal "column" — reaches the column store, and
// every other value including misspellings and different casings reaches the
// row store. That rule is only implicit in the switch, and it had no test, so
// nothing here would have noticed if the default branch changed direction.
//
// It matters because the other two runtimes disagreed with it: pine-java matched
// with equalsIgnoreCase so "Column" selected the column store, and pine-cpp's
// make_frame fell back to column so "colunm" did too. Row and column stores are
// output-equivalent by design, so the divergence never changed a response byte —
// only the memory and performance profile — which is why no end-to-end channel
// could see it and why this belongs in a unit test.
func TestNewFrameDispatchIsExactMatch(t *testing.T) {
	common := map[string]any{"r": "v"}
	items := []map[string]any{{"id": "a"}}

	t.Run("exact column selects the column store", func(t *testing.T) {
		f := NewFrame(StorageModeColumn, common, items)
		if _, ok := f.(*RowFrame); ok {
			t.Fatalf("storage_mode %q must not select the row store", StorageModeColumn)
		}
	})

	// Every other value, including the two that diverged in #179.
	rowModes := []StorageMode{
		StorageModeRow,
		"",
		"colunm",
		"Column",
		"COLUMN",
		"cOlUmN",
		"column ",
		" column",
		"columns",
		"col",
		"rows",
		"unknown",
	}
	for _, mode := range rowModes {
		t.Run("row store for "+string(mode), func(t *testing.T) {
			f := NewFrame(mode, common, items)
			if _, ok := f.(*RowFrame); !ok {
				t.Fatalf("storage_mode %q must select the row store, got %T", mode, f)
			}
			if got := f.ItemCount(); got != 1 {
				t.Fatalf("item count = %d, want 1", got)
			}
		})
	}
}
